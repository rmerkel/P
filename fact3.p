{ Calculate 11 (0..10) factorials
{	 n		    n!	}
{  ---  ---------	}
{	 0		    0	}
{	 1		    1	}
{	 2		    2	}
{	 3		    6	}
{	...				}
{	 9	  362,880	}
{	10	3,628,800	}

const nFacts = 10;
var result;
function factorial(n)
	var p;

	begin
		p = 1;
		while n > 0 do begin
			p = p * n;
			n = n - 1
		end;
		factorial = p;
	end;

begin
	{ call with wrong number of parameters! }
    result = factorial(nFacts, nFacts)
end.
