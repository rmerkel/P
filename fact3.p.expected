./pl0c: passing 2 parameters where 1 where expected near line 29
# fact3.p, 2: { Calculate 11 (0..10) factorials
# fact3.p, 3: {	 n		    n!	}
# fact3.p, 4: {  ---  ---------	}
# fact3.p, 5: {	 0		    0	}
# fact3.p, 6: {	 1		    1	}
# fact3.p, 7: {	 2		    2	}
# fact3.p, 8: {	 3		    6	}
# fact3.p, 9: {	...				}
# fact3.p, 10: {	 9	  362,880	}
# fact3.p, 11: {	10	3,628,800	}
# fact3.p, 12: 
# fact3.p, 13: const nFacts = 10;
    0: call 0, 30
    1: halt
# fact3.p, 14: var result : integer;
# fact3.p, 15: function factorial(n : integer) : integer
# fact3.p, 16: 	var p : integer;
# fact3.p, 17: 
# fact3.p, 18: 	begin
    2: enter 1
# fact3.p, 19: 		p = 1;
    3: pushConst 1
    4: pushVar 0, 4
    5: assign
# fact3.p, 20: 		while n > 0 do begin
    6: pushVar 0, -1
    7: eval
    8: pushConst 0
    9: gt
   10: jneq 25
# fact3.p, 21: 			p = p * n;
   11: pushVar 0, 4
   12: eval
   13: pushVar 0, -1
   14: eval
   15: muli
   16: pushVar 0, 4
   17: assign
# fact3.p, 22: 			n = n - 1
   18: pushVar 0, -1
   19: eval
   20: pushConst 1
# fact3.p, 23: 		end;
   21: subi
   22: pushVar 0, -1
   23: assign
   24: jump 6
# fact3.p, 24: 		factorial = p;
   25: pushVar 0, 4
   26: eval
   27: pushVar 0, 3
   28: assign
# fact3.p, 25: 	end;
   29: reti
# fact3.p, 26: 
# fact3.p, 27: begin
   30: enter 1
# fact3.p, 28: 	{ call with wrong number of parameters! }
# fact3.p, 29:     result = factorial(nFacts, nFacts)
   31: pushConst 10
   32: pushConst 10
# fact3.p, 30: end.
   33: call 0, 2
   34: pushVar 0, 4
   35: assign
   36: ret

